### DISCUSSION TITLE
Оставить в истории прослушиваний Spotify только то, что есть в истории Lastfm

### DISCUSSION THREAD

**Discussion by efremych** [Chimildic/goofy#7] [Помощь с алгоритмом] [answered]
Ну поехали.
Цель: использовать историю Спотифая за минусом треков "на паузе" _(пропуски треков пока цели учитывать нет)_

При текущей реализации:
1. Когда я включу **ON_LASTFM_RECENT_TRACKS** в кэш попадет вся накопившаяся история с Ласт ФМ?
2. Мне надо будет переделать **updateSpotifyRecentTracks** под свои нужды?
3. Когда я переделаю **updateSpotifyRecentTracks** из всего кэша "**BothRecentTracks.json**" уйдут лишние треки?

Или может какой иной вариант, не напрягающий запросами соответствующий сервис?

**Comment by Chimildic** [comment, marked as answer]
> Оставить в выборке Спотифая только те треки, которые присутствуют в выборке Ласта

[removeTracks](https://chimildic.github.io/goofy/#/func?id=removetracks) c `invert = true`
```js
Filter.removeTracks(spotifyTracks, lastfmTracks, true);
```
Если не заводить отдельную функцию/триггер под это, то в `updateBothSourceRecentTracks` рядом с фильтром удаления дубликатов поставить. И через get забирать. 
```js
let tracks = RecentTracks.get();
```

## Старый ответ
Для начала, что такое треки "на паузе"? Текущий играющий/активный трек не возвращается Spotify. Только когда переключишься на следующий. И то не факт. Не смотря на то, что в документации API написано 30+ секунд. Трек может не засчитаться или придти с задержкой. 
В истории ластфм играющий трек тоже игнорируется. Через параметр `ON_LASTFM_RECENT_TRACKS` будут проверяться последние треки истории в количестве `LASTFM_RANGE_RECENT_TRACKS`. Если они новые, добавятся в кэш-файл.

**Reply by efremych** [reply]
"На паузе" - например, я случайно нажал на трек, не прослушивал его, пошел по делам, он висит в плеере выключенным на нулевой-первой секунде. В результате этот трек попадет в историю прослушивания, так как некоторое время висел в плеере.
В Ласт ФМ такие треки не попадают.
Вот и думаю брать за основу Спотифай и просеивать сквозь Ластфм (пропуски треков от ластфм также не нужны пока что).

**Reply by Chimildic** [reply]
Для такой задачи достаточно брать историю с ластфм и отключить spotify. 
```js
// config.gs
UserProperties.setProperty('ON_SPOTIFY_RECENT_TRACKS', 'false');
UserProperties.setProperty('ON_LASTFM_RECENT_TRACKS', 'true');
```
И забирать накопленное через `RecentTracks`
```js
let tracks = RecentTracks.get();
```

**Reply by efremych** [reply]
Так в том то и дело, что меня устраивает "корявость" Спотифая в плане того, что засчитывается только полностью прослушанный трек. А Ласт ФМ засчитывает пропуски - они мне в истории не нужны.

Т.е. как бы я делал без кэша:
1. Взять историю Спотифая - 1000 треков
2. Взять историю Ласта - 1000 треков
3. С допущением что по датам эти 1000 треков могут не биться (у каждого плюс-минус своя история)
4. Оставить в выборке Спотифая только те треки, которые присутствуют в выборке Ласта

Эти бы действия и прописал в updateSpotifyRecentTracks (уже updateBothSourceRecentTracks, точнее), например.
Пока к этому и склоняюсь (предварительно перетащив имеющуюся историю Ласт фм в кэш LastfmRecentTracks.json)

**Reply by Chimildic** [reply]
> Оставить в выборке Спотифая только те треки, которые присутствуют в выборке Ласта

[removeTracks](https://chimildic.github.io/goofy/#/func?id=removetracks) c `invert = true`
```js
Filter.removeTracks(spotifyTracks, lastfmTracks, true);
```
Если не заводить отдельную функцию/триггер под это, то в `updateBothSourceRecentTracks` рядом с фильтром удаления дубликатов поставить. И через get забирать. 
```js
let tracks = RecentTracks.get();
```

**Comment by efremych** [comment]
А почему данный код не сжимает треки?
Есть файл с накопленными данными от ЛастФМ, хочу уменьшить его объем, как полагается

 let tracks = Source.getPlaylistTracks('playlist name', '3ihbk4jhoI602T5S3AJZ0r');
Cache.compressTracks(tracks);
Cache.write('myfile.json', tracks);

**Reply by Chimildic** [reply]
Потому что Spotify решил, что удачно одинаковым ключам давать разные типы в разных запросах. Поправил, обновитесь или измените строчку в `compressTracks`. Лучше обновиться, были и другие правки. 
```js
        tracks.forEach((item) => {
            if (typeof item.track === 'object') { // эту
                // ...
            }
            // ...
        });
```

Для сжатия существующих файлов истории лучше использовать `RecentTracks.compress`

**Reply by efremych** [reply]
а у вас removeTracks работает, если использовать эту функцию для двух кэшей? Если работает, буду у себя искать...

let tracks = Cache.read('SpotifyRecentTracks.json');
let lastfm = Cache.read('LastfmRecentTracks.json');
Filter.removeTracks(tracks, lastfm);

На getTrackKey(track) жалуется.

**Reply by Chimildic** [reply]
Внутри `SpotifyRecentTracks.json` другая структура элементов. Чтобы прочитать треки `Source.getRecentTracks`. 

Зачем это делать? Я же написал куда встроить `removeTracks`
> в `updateBothSourceRecentTracks` рядом с фильтром удаления дубликатов поставить. И через get забирать.

**Reply by efremych** [reply]
ок, ясно.
Теперь не удается конфиг запустить :)
ReferenceError: userProperties is not defined (line 3, file "!config")

Всё обновил, вроде бы удачно

**Reply by efremych** [reply]
Или я что не так делаю?

Меняю на тру в конфиге и нажимаю запуск
    userProperties.setProperty('ON_LASTFM_RECENT_TRACKS', 'false');

**Reply by Chimildic** [reply]
Обновите `config.gs` до [актуального состояния](https://github.com/Chimildic/goofy/blob/main/config.js)

**Reply by efremych** [reply]
Вроде то же самое, различий не нвижу

```
function setProperties(){
// Описание параметров: https://github.com/Chimildic/goofy/blob/main/guide/README.md#параметры
    userProperties.setProperty('CLIENT_ID', 'айди');
    userProperties.setProperty('CLIENT_SECRET', 'айди');
    userProperties.setProperty('LASTFM_API_KEY', 'айди');
    userProperties.setProperty('ON_SPOTIFY_RECENT_TRACKS', 'true');
    userProperties.setProperty('ON_LASTFM_RECENT_TRACKS', 'false');
    userProperties.setProperty('LASTFM_RANGE_RECENT_TRACKS', '30');
    userProperties.setProperty('LASTFM_LOGIN', 'логин');
}

function logProperties(){
    console.log(UserProperties.getProperties());
}
```

**Reply by efremych** [reply]
Библиотека

```
const VERSION = '1.3.1';
const UserProperties = PropertiesService.getUserProperties();
const KeyValue = UserProperties.getProperties();
const CLIENT_ID = KeyValue.CLIENT_ID;
const CLIENT_SECRET = KeyValue.CLIENT_SECRET;
const LASTFM_API_KEY = KeyValue.LASTFM_API_KEY;
const ON_SPOTIFY_RECENT_TRACKS = 'true' === KeyValue.ON_SPOTIFY_RECENT_TRACKS;
const ON_LASTFM_RECENT_TRACKS = 'true' === KeyValue.ON_LASTFM_RECENT_TRACKS;
const LASTFM_RANGE_RECENT_TRACKS = parseInt(KeyValue.LASTFM_RANGE_RECENT_TRACKS);
const LASTFM_LOGIN = KeyValue.LASTFM_LOGIN;
const API_BASE_URL = 'https://api.spotify.com/v1';
```

**Reply by Chimildic** [reply]
`userProperties` должно быть с большой буквы `UserProperties`

**Comment by efremych** [comment]
Спасибо большое!
Сам себе усложнил жизнь, пытаясь из плейлиста Историю сваять (не спрашивайте зачем). В итоге потерял часть истории (из копии восстановил большую часть).

Но вроде, что хотел получил - будем наблюдать.
В итоге вот так (ну, как изначально и писалось))). Удаление дубликатов убрал, чтобы был плейлист с реальной историей.
```
    function updateBothSourceRecentTracks() {
        let spotifyTracks = Source.getRecentTracks(ITEMS_LIMIT);
        let lastfmTracks = Cache.read(LASTFM_FILENAME);
        Filter.removeTracks(spotifyTracks, lastfmTracks, invert = true);
//        Combiner.push(spotifyTracks, lastfmTracks);
//        Filter.dedupTracks(spotifyTracks);
        spotifyTracks.sort((x, y) => Order.compareDate(y.played_at, x.played_at));
        Cache.write(BOTH_SOURCE_FILENAME, spotifyTracks);
    }
```
Время выполнения тригера правда увеличилось в несколько раз (учитывая, что вся возня ради исключения нескольких треков в неделю)))). Ну, надеюсь, не критично.
![image](https://user-images.githubusercontent.com/76212904/102698259-24aa7d00-424d-11eb-9875-eca92952df82.png)
